File: hdf-converters/shared/typescript/bom/fingerprints.ts
/**
 * BOM format fingerprints + detectFormat dispatcher.
 *
 * Self-contained structural detectors for the three supported BOM formats.
 * Detection precedence: an ML-BOM is a CycloneDX document with a
 * machine-learning-model component, so the ML detector must win over the plain
 * CycloneDX detector.
 *
 * ASYMMETRY NOTE (parity with Go): the Go peer exposes the same detectFormat
 * logic but has NO fingerprint registry — CLI auto-detect wiring is Phase 3 in
 * both languages. These BOM detectors are deliberately NOT registered in the
 * shared ConverterFingerprint registry (shared/typescript/registry.ts): that
 * registry drives full-document CLI auto-detection, and a BOM detector
 * returning 1.0 for a CycloneDX file would tie with the existing
 * cyclonedx-to-hdf converter fingerprint, making detectConverter refuse to
 * guess (ambiguous-tie rule) and regressing every CycloneDX input. BOM parsing
 * is a sub-document concern with no CLI converter until Phase 3.
 */

import type { BomFormat } from './model.js';

export interface FormatDetection {
  format: BomFormat;
  /** Detection confidence in the range 0..1. */
  confidence: number;
}

function record(input: unknown): Record<string, unknown> | undefined {
  return typeof input === 'object' && input !== null && !Array.isArray(input)
    ? (input as Record<string, unknown>)
    : undefined;
}

/** CycloneDX: bomFormat === 'CycloneDX'. Returns 0..1 confidence. */
export function detectCycloneDX(input: unknown): number {
  const obj = record(input);
  return obj?.bomFormat === 'CycloneDX' ? 1 : 0;
}

/**
 * CycloneDX ML-BOM: a CycloneDX document with at least one
 * machine-learning-model component. Strictly more specific than plain
 * CycloneDX.
 */
export function detectCycloneDXML(input: unknown): number {
  const obj = record(input);
  if (obj?.bomFormat !== 'CycloneDX') return 0;
  const components = obj.components;
  if (!Array.isArray(components)) return 0;
  const hasModel = components.some(
    c => record(c)?.type === 'machine-learning-model',
  );
  return hasModel ? 1 : 0;
}

/** SPDX: a non-empty spdxVersion string is present. Returns 0..1 confidence. */
export function detectSPDX(input: unknown): number {
  const obj = record(input);
  return typeof obj?.spdxVersion === 'string' && obj.spdxVersion.length > 0 ? 1 : 0;
}

/**
 * SPDX 3.0 AI/Dataset (JSON-LD): a document with an `@context` and an `@graph`
 * array carrying at least one ai_AIPackage or dataset_DatasetPackage element.
 * Structurally disjoint from the SPDX 2.3 detector (which keys on spdxVersion),
 * so the two never conflict.
 */
export function detectSPDX3(input: unknown): number {
  const obj = record(input);
  if (!obj || obj['@context'] === undefined) return 0;
  const graph = obj['@graph'];
  if (!Array.isArray(graph)) return 0;
  const hasAISubject = graph.some(el => {
    const type = record(el)?.type;
    return type === 'ai_AIPackage' || type === 'dataset_DatasetPackage';
  });
  return hasAISubject ? 1 : 0;
}

/**
 * Detect the BOM format of a parsed JSON object. ML wins over plain CycloneDX
 * by precedence; returns undefined when no supported format matches.
 */
export function detectFormat(input: unknown): FormatDetection | undefined {
  const ml = detectCycloneDXML(input);
  if (ml > 0) return { format: 'cyclonedx-ml', confidence: ml };
  const cdx = detectCycloneDX(input);
  if (cdx > 0) return { format: 'cyclonedx', confidence: cdx };
  const spdx3 = detectSPDX3(input);
  if (spdx3 > 0) return { format: 'spdx-3-ai', confidence: spdx3 };
  const spdx = detectSPDX(input);
  if (spdx > 0) return { format: 'spdx', confidence: spdx };
  return undefined;
}

File: hdf-converters/shared/go/bom/fingerprints.go
// BOM format fingerprints + DetectFormat dispatcher.
//
// Self-contained structural detectors for the three supported BOM formats.
// Detection precedence: an ML-BOM is a CycloneDX document with a
// machine-learning-model component, so the ML detector must win over the plain
// CycloneDX detector.

package bom

// FormatDetection reports a detected BOM format and a 0..1 confidence.
type FormatDetection struct {
	Format     string
	Confidence float64
}

// DetectCycloneDX reports confidence that input is a CycloneDX document
// (bomFormat === "CycloneDX"). Returns 0 or 1.
func DetectCycloneDX(input any) float64 {
	obj := asRecord(input)
	if obj != nil && obj["bomFormat"] == "CycloneDX" {
		return 1
	}
	return 0
}

// DetectCycloneDXML reports confidence that input is a CycloneDX ML-BOM: a
// CycloneDX document with at least one machine-learning-model component.
// Strictly more specific than plain CycloneDX. Returns 0 or 1.
func DetectCycloneDXML(input any) float64 {
	obj := asRecord(input)
	if obj == nil || obj["bomFormat"] != "CycloneDX" {
		return 0
	}
	components, ok := obj["components"].([]any)
	if !ok {
		return 0
	}
	for _, c := range components {
		if asRecord(c)["type"] == "machine-learning-model" {
			return 1
		}
	}
	return 0
}

// DetectSPDX reports confidence that input is an SPDX document (a non-empty
// spdxVersion string is present). Returns 0 or 1.
func DetectSPDX(input any) float64 {
	obj := asRecord(input)
	if obj != nil && asString(obj["spdxVersion"]) != "" {
		return 1
	}
	return 0
}

// DetectSPDX3 reports confidence that input is an SPDX 3.0 AI/Dataset document
// (JSON-LD): an `@context` plus an `@graph` array carrying at least one
// ai_AIPackage or dataset_DatasetPackage element. Structurally disjoint from the
// SPDX 2.3 detector (which keys on spdxVersion), so the two never conflict.
// Returns 0 or 1.
func DetectSPDX3(input any) float64 {
	obj := asRecord(input)
	if obj == nil {
		return 0
	}
	if _, ok := obj["@context"]; !ok {
		return 0
	}
	graph, ok := obj["@graph"].([]any)
	if !ok {
		return 0
	}
	for _, el := range graph {
		t := asString(asRecord(el)["type"])
		if t == "ai_AIPackage" || t == "dataset_DatasetPackage" {
			return 1
		}
	}
	return 0
}

// DetectFormat detects the BOM format of a parsed JSON value. ML wins over plain
// CycloneDX by precedence; returns nil when no supported format matches.
func DetectFormat(input any) *FormatDetection {
	if ml := DetectCycloneDXML(input); ml > 0 {
		return &FormatDetection{Format: FormatCycloneDXML, Confidence: ml}
	}
	if cdx := DetectCycloneDX(input); cdx > 0 {
		return &FormatDetection{Format: FormatCycloneDX, Confidence: cdx}
	}
	if spdx3 := DetectSPDX3(input); spdx3 > 0 {
		return &FormatDetection{Format: FormatSPDX3AI, Confidence: spdx3}
	}
	if spdx := DetectSPDX(input); spdx > 0 {
		return &FormatDetection{Format: FormatSPDX, Confidence: spdx}
	}
	return nil
}
